src/LowerParallelTasks.cpp

```diff
@@ -0,0 +1,519 @@
+                result = Call::make(Int(32), "halide_do_par_for", args, Call::Extern);
+            } else {
+                std::vector<Expr> function_decl_args(5);
+                function_decl_args[0] = make_zero(type_of<void *>());
+                function_decl_args[1] = make_zero(Int(32));
+                function_decl_args[2] = make_zero(Int(32));
+                function_decl_args[3] = make_zero(type_of<uint8_t *>());
+                function_decl_args[4] = make_zero(type_of<void *>());
+                Expr function_decl_call = Call::make(Int(32), new_function_name, function_decl_args, Call::Extern);
+
+                tasks_array_args.emplace_back(Call::make(type_of<const void *>(), Call::resolve_function_name, {function_decl_call}, Call::PureIntrinsic));
+                tasks_array_args.emplace_back(Cast::make(type_of<uint8_t *>(), closure_struct));
+                tasks_array_args.emplace_back(StringImm::make(t.name));
+                tasks_array_args.emplace_back(semaphores_array.defined() ? semaphores_array : semaphore_type);
+                tasks_array_args.emplace_back((int)t.semaphores.size());
+                tasks_array_args.emplace_back(t.min);
+                tasks_array_args.emplace_back(t.extent);
+                tasks_array_args.emplace_back(min_threads.result);
+                tasks_array_args.emplace_back(Cast::make(Bool(), t.serial));
+            }
+        }
+
+        if (tasks_array_args.size() > 2) {
+            // Allocate task list array
+            Expr tasks_list = Call::make(Handle(), Call::make_typed_struct, tasks_array_args, Call::PureIntrinsic);
+            Expr user_context = Call::make(type_of<void *>(), Call::get_user_context, {}, Call::PureIntrinsic);
+            Expr task_parent = Call::make(Handle(), Call::get_pointer_symbol_or_null,
+                                          {StringImm::make("_task_parent"), make_zero(Handle())}, Call::Intrinsic);
+            result = Call::make(Int(32), "halide_do_parallel_tasks",
+                                {user_context, make_const(Int(32), num_tasks), tasks_list, task_parent},
+                                Call::Extern);
+        }
+
+        result = Let::make(closure_name, closure_struct_allocation, result);
+        std::string closure_result_name = unique_name("closure_result");
+        Expr closure_result = Variable::make(Int(32), closure_result_name);
+        return LetStmt::make(closure_result_name, result, AssertStmt::make(closure_result == 0, closure_result));
+    }
+
+    void get_parallel_tasks(const Stmt &s, std::vector<ParallelTask> &result, std::pair<std::string, int> prefix) {
+        const For *loop = s.as<For>();
+        const Acquire *acquire = loop ? loop->body.as<Acquire>() : s.as<Acquire>();
+        if (const Fork *f = s.as<Fork>()) {
+            add_fork(prefix);
+            get_parallel_tasks(f->first, result, prefix);
+            get_parallel_tasks(f->rest, result, prefix);
```

Comment:
Do Forks ever get nested deeply enough that we should worry about recursion here?